File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/java/src/main/java/com/ondc/ONDCSignature.java
package com.ondc;

import org.bouncycastle.crypto.digests.Blake2bDigest;
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters;
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters;
import org.bouncycastle.crypto.signers.Ed25519Signer;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;
import java.util.TreeMap;

/**
 * ONDC Ed25519 Signature Generator - Java Implementation
 * 
 * Implements the ONDC protocol signature scheme for V3 API authentication.
 */
public class ONDCSignature {
    private final String subscriberId;
    private final String uniqueKeyId;
    private final Ed25519PrivateKeyParameters privateKey;
    private final Ed25519PublicKeyParameters publicKey;
    private final String publicKeyB64;
    private final Gson gson;
    
    /**
     * Initialize ONDC signature generator.
     *
     * @param subscriberId   Unique subscriber ID
     * @param uniqueKeyId    Unique key identifier
     * @param privateKeySeed 32-byte private key seed (hex string)
     */
    public ONDCSignature(String subscriberId, String uniqueKeyId, String privateKeySeedHex) {
        this.subscriberId = subscriberId;
        this.uniqueKeyId = uniqueKeyId;
        
        // Convert hex string to bytes
        byte[] seed = hexStringToByteArray(privateKeySeedHex);
        if (seed.length != 32) {
            throw new IllegalArgumentException("Private key seed must be exactly 32 bytes");
        }
        
        // Generate Ed25519 key pair from seed
        this.privateKey = new Ed25519PrivateKeyParameters(seed, 0);
        this.publicKey = privateKey.generatePublicKey();
        this.publicKeyB64 = getPublicKeyB64();
        
        // Initialize Gson with specific formatting for consistent JSON
        this.gson = new GsonBuilder()
            .disableHtmlEscaping()
            .serializeNulls()
            .create();
    }
    
    /**
     * Get DER/SPKI-encoded public key (base64).
     */
    private String getPublicKeyB64() {
        byte[] rawPubKey = publicKey.getEncoded();
        
        // Wrap in SPKI/DER format (matching Python's SubjectPublicKeyInfo)
        byte[] spkiHeader = new byte[]{
            0x30, 0x2a,  // SEQUENCE, 42 bytes
            0x30, 0x05,  // SEQUENCE, 5 bytes
            0x06, 0x03, 0x2b, 0x65, 0x70,  // OID 1.3.101.112 (Ed25519)
            0x03, 0x21, 0x00  // BIT STRING, 33 bytes (including 0x00 padding)
        };
        
        byte[] spkiEncoded = new byte[spkiHeader.length + rawPubKey.length];
        System.arraycopy(spkiHeader, 0, spkiEncoded, 0, spkiHeader.length);
        System.arraycopy(rawPubKey, 0, spkiEncoded, spkiHeader.length, rawPubKey.length);
        
        return Base64.getEncoder().encodeToString(spkiEncoded);
    }
    
    /**
     * Create BLAKE2b-512 digest of request body.
     */
    private String createDigest(Object body) {
        byte[] bodyBytes;
        
        if (body == null) {
            bodyBytes = new byte[0];
        } else if (body instanceof String) {
            bodyBytes = ((String) body).getBytes(StandardCharsets.UTF_8);
        } else if (body instanceof Map) {
            // Convert to TreeMap to sort keys alphabetically for consistency
            Map<String, Object> sortedBody = new TreeMap<>((Map<String, Object>) body);
            // Recursively sort nested maps
            sortedBody = sortMapRecursively(sortedBody);
            
            // Serialize with Gson and format to match Python's json.dumps(separators=(', ', ': '))
            String json = gson.toJson(sortedBody);
            // Add spaces after JSON structure separators only (not within string values)
            json = json.replaceAll("\":", "\": ").replaceAll(",\"", ", \"").replaceAll("\\},\\{" , "}, {");
            bodyBytes = json.getBytes(StandardCharsets.UTF_8);
        } else {
            throw new IllegalArgumentException("Body must be null, String, or Map");
        }
        
        // Generate BLAKE2b-512 hash
        Blake2bDigest digest = new Blake2bDigest(512);
        digest.update(bodyBytes, 0, bodyBytes.length);
        byte[] hash = new byte[64];
        digest.doFinal(hash, 0);
        
        return Base64.getEncoder().encodeToString(hash);
    }
    
    /**
     * Create the signing string according to ONDC spec.
     */
    private String createSigningString(long created, long expires, String digest) {
        return String.format("(created): %d\n(expires): %d\ndigest: BLAKE-512=%s",
                           created, expires, digest);
    }
    
    /**
     * Sign the signing string with Ed25519 private key.
     */
    private String signString(String signingString) {
        Ed25519Signer signer = new Ed25519Signer();
        signer.init(true, privateKey);
        
        byte[] message = signingString.getBytes(StandardCharsets.UTF_8);
        signer.update(message, 0, message.length);
        byte[] signature = signer.generateSignature();
        
        return Base64.getEncoder().encodeToString(signature);
    }
    
    /**
     * Generate complete Signature authorization header.
     *
     * @param body    Request body (can be null, String, or Map)
     * @param created Unix timestamp when signature was created (null for current time)
     * @param expires Unix timestamp when signature expires (null for created + 300)
     * @return Authorization header string
     */
    public String generateSignatureHeader(Object body, Long created, Long expires) {
        // Generate timestamps
        if (created == null) {
            created = System.currentTimeMillis() / 1000;
        }
        if (expires == null) {
            // 60-second validity window (ONDC recommended: 30-60 seconds)
            expires = created + 60;
        }
        
        // Create digest
        String digest = createDigest(body);
        
        // Create signing string
        String signingString = createSigningString(created, expires, digest);
        
        // Generate signature
        String signature = signString(signingString);
        
        // Construct Signature header
        String keyId = String.format("%s|%s|ed25519", subscriberId, uniqueKeyId);
        
        return String.format(
            "Signature keyId=\"%s\",algorithm=\"ed25519\",created=\"%d\",expires=\"%d\",headers=\"(created) (expires) digest\",signature=\"%s\"",
            keyId, created, expires, signature
        );
    }
    
    /**
     * Generate Digest header.
     */
    public String generateDigestHeader(Object body) {
        return "BLAKE-512=" + createDigest(body);
    }
    
    /**
     * Get public key (base64).
     */
    public String getPublicKey() {
        return publicKeyB64;
    }
    
    /**
     * Serialize body to JSON string matching the format used in digest calculation.
     * Use this when sending HTTP requests to ensure digest matches.
     */
    public String serializeBody(Object body) {
        if (body == null || (body instanceof String && ((String) body).isEmpty())) {
            return "";
        } else if (body instanceof String) {
            return (String) body;
        } else if (body instanceof Map) {
            Map<String, Object> sortedBody = sortMapRecursively((Map<String, Object>) body);
            // Serialize with Gson and format to match Python's json.dumps(separators=(', ', ': '))
            String json = gson.toJson(sortedBody);
            // Add spaces after JSON structure separators only (not within string values)
            json = json.replaceAll("\":", "\": ").replaceAll(",\"", ", \"").replaceAll("\\},\\{", "}, {");
            return json;
        }
        return gson.toJson(body);
    }
    
    /**
     * Helper: Convert hex string to byte array.
     */
    private static byte[] hexStringToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                                + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }
    
    /**
     * Helper: Recursively sort maps to ensure consistent key ordering.
     */
    @SuppressWarnings("unchecked")
    private Map<String, Object> sortMapRecursively(Map<String, Object> map) {
        Map<String, Object> sorted = new TreeMap<>();
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            Object value = entry.getValue();
            if (value instanceof Map) {
                sorted.put(entry.getKey(), sortMapRecursively((Map<String, Object>) value));
            } else if (value instanceof java.util.List) {
                // Handle arrays - recursively sort maps inside array elements
                java.util.List<?> list = (java.util.List<?>) value;
                java.util.List<Object> sortedList = new java.util.ArrayList<>();
                for (Object item : list) {
                    if (item instanceof Map) {
                        sortedList.add(sortMapRecursively((Map<String, Object>) item));
                    } else {
                        sortedList.add(item);
                    }
                }
                sorted.put(entry.getKey(), sortedList);
            } else {
                sorted.put(entry.getKey(), value);
            }
        }
        return sorted;
    }
}

File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/golang/ondc_signature.go
// ONDC Ed25519 Signature Generator - Go Implementation
//
// Implements the ONDC protocol signature scheme for V3 API authentication.

package main

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	"golang.org/x/crypto/blake2b"
)

// ONDCSignature handles ONDC signature generation
type ONDCSignature struct {
	SubscriberID string
	UniqueKeyID  string
	PrivateKey   ed25519.PrivateKey
	PublicKey    ed25519.PublicKey
	PublicKeyB64 string
}

// NewONDCSignature creates a new ONDC signature generator
func NewONDCSignature(subscriberID, uniqueKeyID, privateKeySeedHex string) (*ONDCSignature, error) {
	// Convert hex string to bytes
	seed, err := hex.DecodeString(privateKeySeedHex)
	if err != nil {
		return nil, fmt.Errorf("invalid hex seed: %w", err)
	}

	if len(seed) != 32 {
		return nil, fmt.Errorf("private key seed must be exactly 32 bytes, got %d", len(seed))
	}

	// Generate Ed25519 key pair from seed
	privateKey := ed25519.NewKeyFromSeed(seed)
	publicKey := privateKey.Public().(ed25519.PublicKey)

	o := &ONDCSignature{
		SubscriberID: subscriberID,
		UniqueKeyID:  uniqueKeyID,
		PrivateKey:   privateKey,
		PublicKey:    publicKey,
	}

	o.PublicKeyB64 = o.getPublicKeyB64()

	return o, nil
}

// getPublicKeyB64 returns DER/SPKI-encoded public key (base64)
func (o *ONDCSignature) getPublicKeyB64() string {
	// SPKI/DER header for Ed25519 public key
	spkiHeader := []byte{
		0x30, 0x2a, // SEQUENCE, 42 bytes
		0x30, 0x05, // SEQUENCE, 5 bytes
		0x06, 0x03, 0x2b, 0x65, 0x70, // OID 1.3.101.112 (Ed25519)
		0x03, 0x21, 0x00, // BIT STRING, 33 bytes (including 0x00 padding)
	}

	spkiEncoded := append(spkiHeader, o.PublicKey...)
	return base64.StdEncoding.EncodeToString(spkiEncoded)
}

// serializeJSON converts a body to JSON string with consistent formatting
func (o *ONDCSignature) serializeJSON(body interface{}) (string, error) {
	// Sort the object keys for consistent serialization
	sortedBody := sortMapKeys(body)

	// Marshal object to JSON with SetEscapeHTML(false) to match Python's ensure_ascii=False
	buf := new(strings.Builder)
	encoder := json.NewEncoder(buf)
	encoder.SetEscapeHTML(false)
	err := encoder.Encode(sortedBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal body: %w", err)
	}

	// Remove trailing newline added by Encoder
	jsonStr := strings.TrimSuffix(buf.String(), "\n")

	// Add spaces after colons and commas to match Python format
	jsonStr = strings.ReplaceAll(jsonStr, "\":", "\": ")
	jsonStr = strings.ReplaceAll(jsonStr, ",\"", ", \"")
	jsonStr = strings.ReplaceAll(jsonStr, "},{", "}, {")

	return jsonStr, nil
}

// createDigest creates BLAKE2b-512 digest of request body
func (o *ONDCSignature) createDigest(body interface{}) (string, error) {
	var bodyBytes []byte

	if body == nil {
		bodyBytes = []byte{}
	} else if str, ok := body.(string); ok {
		bodyBytes = []byte(str)
	} else if b, ok := body.([]byte); ok {
		bodyBytes = b
	} else {
		jsonStr, err := o.serializeJSON(body)
		if err != nil {
			return "", err
		}
		bodyBytes = []byte(jsonStr)
	}

	// Generate BLAKE2b-512 hash
	hash := blake2b.Sum512(bodyBytes)
	return base64.StdEncoding.EncodeToString(hash[:]), nil
}

// createSigningString creates the signing string according to ONDC spec
func (o *ONDCSignature) createSigningString(created, expires int64, digest string) string {
	return fmt.Sprintf("(created): %d\n(expires): %d\ndigest: BLAKE-512=%s", created, expires, digest)
}

// signString signs the signing string with Ed25519 private key
func (o *ONDCSignature) signString(signingString string) string {
	signature := ed25519.Sign(o.PrivateKey, []byte(signingString))
	return base64.StdEncoding.EncodeToString(signature)
}

// GenerateSignatureHeader generates complete Signature authorization header
func (o *ONDCSignature) GenerateSignatureHeader(body interface{}, created, expires *int64) (string, string, error) {
	// Generate timestamps
	var createdTs, expiresTs int64
	if created == nil {
		createdTs = getCurrentTimestamp()
	} else {
		createdTs = *created
	}
	if expires == nil {
		// 60-second validity window (ONDC recommended: 30-60 seconds)
		expiresTs = createdTs + 60
	} else {
		expiresTs = *expires
	}

	// Create digest
	digest, err := o.createDigest(body)
	if err != nil {
		return "", "", err
	}
	digestHeader := fmt.Sprintf("BLAKE-512=%s", digest)

	// Create signing string
	signingString := o.createSigningString(createdTs, expiresTs, digest)

	// Generate signature
	signature := o.signString(signingString)

	// Construct Signature header
	keyID := fmt.Sprintf("%s|%s|ed25519", o.SubscriberID, o.UniqueKeyID)

	authHeader := fmt.Sprintf(
		`Signature keyId="%s",algorithm="ed25519",created="%d",expires="%d",headers="(created) (expires) digest",signature="%s"`,
		keyID, createdTs, expiresTs, signature,
	)

	return authHeader, digestHeader, nil
}

// GetPublicKey returns the public key (base64)
func (o *ONDCSignature) GetPublicKey() string {
	return o.PublicKeyB64
}

// SerializeBody serializes body to JSON string matching the format used in digest calculation
// Use this when sending HTTP requests to ensure digest matches
func (o *ONDCSignature) SerializeBody(body interface{}) (string, error) {
	if body == nil {
		return "", nil
	} else if str, ok := body.(string); ok {
		return str, nil
	} else {
		return o.serializeJSON(body)
	}
}

// Helper function to get current Unix timestamp
func getCurrentTimestamp() int64 {
	return time.Now().Unix()
}

// sortMapKeys recursively sorts map keys alphabetically for consistent JSON serialization
func sortMapKeys(obj interface{}) interface{} {
	switch v := obj.(type) {
	case map[string]interface{}:
		// Get all keys and sort them
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		// Create new map with sorted keys
		sorted := make(map[string]interface{})
		for _, k := range keys {
			sorted[k] = sortMapKeys(v[k])
		}
		return sorted
	case []interface{}:
		// Recursively sort arrays
		result := make([]interface{}, len(v))
		for i, item := range v {
			result[i] = sortMapKeys(item)
		}
		return result
	default:
		return obj
	}
}
